File: iosApp/iosApp/Presentation/Main/Admin/User/UserManagementViewModel.swift
import Foundation
import Combine
import Shared

@MainActor
final class UserManagementViewModel: ObservableObject {
    private let getAdminUserPageUseCase: GetAdminUserPageUseCase

    @Published private(set) var uiState = UserManagementUiState()

    let event = PassthroughSubject<UserManagementEvent, Never>()

    private var pageTask: Task<Void, Never>?

    func onAction(_ action: UserManagementAction) {
        switch action {
        case .selectFilter(let filter):
            guard uiState.selectedFilter != filter else { return }
            uiState.selectedFilter = filter
            uiState.users = []
            uiState.nextCursor = nil
            uiState.canLoadMore = false
            uiState.infoMessage = nil
            loadPage(cursor: nil, append: false)
        case .loadMore:
            loadMore()
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    private func loadMore() {
        guard uiState.canLoadMore, !uiState.isLoadingMore, let cursor = uiState.nextCursor else { return }
        loadPage(cursor: cursor, append: true)
    }

    private func loadPage(cursor: String?, append: Bool) {
        pageTask?.cancel()
        pageTask = Task { @MainActor in
            if append {
                uiState.isLoadingMore = true
                do {
                    try await Task.sleep(nanoseconds: userPaginationDelayNanoseconds)
                    if Task.isCancelled { return }
                } catch {
                    return
                }
            } else {
                uiState.isLoading = true
                uiState.infoMessage = nil
            }

            do {
                let result = try await getAdminUserPageUseCase.invoke(
                    filter: uiState.selectedFilter,
                    cursor: cursor,
                    pageSize: userManagementPageSize
                )
                if let success = result as? AppResultSuccess<AnyObject>,
                   let page = success.data as? Shared.PagedResult<User> {
                    let pageItems = (page.items as? [User] ?? []).sorted { $0.createdAt > $1.createdAt }
                    uiState.users = append ? (uiState.users + pageItems) : pageItems
                    uiState.nextCursor = page.nextCursor
                    uiState.canLoadMore = page.hasNext
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                    uiState.infoMessage = nil
                } else if let failure = result as? AppResultFailure {
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                    uiState.infoMessage = "\(failure.error)"
                } else {
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.isLoadingMore = false
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    init(
        getAdminUserPageUseCase: GetAdminUserPageUseCase = KoinInitializerKt.resolveGetAdminUserPageUseCase()
    ) {
        self.getAdminUserPageUseCase = getAdminUserPageUseCase

        loadPage(cursor: nil, append: false)
    }

    deinit {
        pageTask?.cancel()
    }
}

private let userManagementPageSize: Int32 = 15
private let userPaginationDelayNanoseconds: UInt64 = 1_000_000_000

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/admin/user/UserManagementViewModel.kt
package com.hhp227.concafe.presentation.main.admin.user

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.usecase.GetAdminUserPageUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class UserManagementViewModel(
    private val getAdminUserPageUseCase: GetAdminUserPageUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(UserManagementUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<UserManagementEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private var pageJob: Job? = null

    fun onAction(action: UserManagementAction) {
        when (action) {
            is UserManagementAction.SelectFilter -> {
                if (_uiState.value.selectedFilter != action.filter) {
                    _uiState.update {
                        it.copy(
                            selectedFilter = action.filter,
                            users = emptyList(),
                            nextCursor = null,
                            canLoadMore = false,
                            infoMessage = null
                        )
                    }
                    loadPage(cursor = null, append = false)
                }
            }
            UserManagementAction.LoadMore -> loadMore()
            UserManagementAction.DismissInfoMessage -> {
                _uiState.update { it.copy(infoMessage = null) }
            }
        }
    }

    private fun loadMore() {
        val state = _uiState.value
        val cursor = state.nextCursor
        if (cursor == null || !state.canLoadMore || state.isLoadingMore) return
        loadPage(cursor = cursor, append = true)
    }

    private fun loadPage(cursor: String?, append: Boolean) {
        pageJob?.cancel()
        pageJob = viewModelScope.launch {
            if (append) {
                _uiState.update { it.copy(isLoadingMore = true) }
                delay(PAGINATION_DELAY_MILLIS)
            } else {
                _uiState.update { it.copy(isLoading = true, infoMessage = null) }
            }

            when (val result = getAdminUserPageUseCase.invoke(
                filter = _uiState.value.selectedFilter,
                cursor = cursor,
                pageSize = USER_PAGE_SIZE
            )) {
                is AppResult.Success -> {
                    val pageItems = result.data.items.sortedByDescending { it.createdAt }
                    _uiState.update { state ->
                        state.copy(
                            users = if (append) state.users + pageItems else pageItems,
                            nextCursor = result.data.nextCursor,
                            canLoadMore = result.data.hasNext,
                            isLoading = false,
                            isLoadingMore = false,
                            infoMessage = null
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            isLoadingMore = false,
                            infoMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    override fun onCleared() {
        pageJob?.cancel()
        super.onCleared()
    }

    init {
        loadPage(cursor = null, append = false)
    }
}

private const val USER_PAGE_SIZE = 15
private const val PAGINATION_DELAY_MILLIS = 1_000L
